### DISCUSSION TITLE
Remaining bugs/limitations of `BaseHTTPMiddleware` - and what to do about them

### DISCUSSION THREAD

**Discussion by jhominal** [encode/starlette#1729] [Q&A]
When #1438 is fixed, the documentation explanation for deprecating `BaseHTTPMiddleware` should be reviewed, in particular its remaining bugs and limitations.

Personally, I can think of two issues in the implementation of `BaseHTTPMiddleware` (once #1438 is fixed):

 1. Context variables that are set downstream cannot go "upstream" of a `BaseHTTPMiddleware` - that is the consequence of the downstream ASGI application being run in a task distinct from the upstream `app`;
 2. If there is a request body and `BaseHTTPMiddleware.dispatch` reads from it, then the request body will not be available for `call_next` - there will be either a deadlock, the first packet will be missing, etc. - that is the consequence of `call_next` hooking itself directly to `request.receive`, and of there being no attempt at buffering the request body (which, in my opinion, would be a bad idea);

Does anyone have any other bugs/limitations in mind?

Now, what to do about the issues I mentioned?

 - For the context variables issue:
   -  I suspect that it can be "mostly fixed" by running `dispatch` in a new task, and running the downstream ASGI app in the same task (no guarantee, it came to my mind as I was writing this post);
   - this has likely tradeoffs around the behavior of context variables, e.g. if the middleware itself sets them;
   - but the whole weight of these tradeoffs would be on the people who are using `BaseHTTPMiddleware`, instead of the current situation where the people who use `BaseHTTPMiddleware` can be completely unaware of the way that they are breaking context variables for other applications and middlewares;
   - I think such a solution would have the advantage that people who currently have `BaseHTTPMiddleware` that do not make use of context variables would have nothing to change, and their middlewares would stop breaking other people's context variables;
   - I could spend a few hours trying that, is anyone interested?
 - For the read request body issue, I would be in favor of keeping that limitation in `BaseHTTPMiddleware`, and document it more strongly, and I would even look into raising an exception if `dispatch` calls both `request.receive` (directly or indirectly) *and* `call_next`;

**Comment by jhominal** [comment]
I would like to add that it is not my intent to question the deprecation of `BaseHTTPMiddleware`, but only to limit the damage that it can do for as long as it keeps being used.

**Reply by adriangb** [reply]
👍 so far I think I am the only one publicly advocating for deprecation, it is certainly not a concensus from the entire team or anything like that. I've also been working both sides of the coin: advocating for deprecation but also trying to fix the issues.  Appreciate your work and approach on this.

**Comment by adriangb** [comment]
> For the context variables issue [...] I suspect that it can be "mostly fixed" by running dispatch in a new task

I'm not as good at visualizing streams/tasks in my head as it seems you are, I think I would need to see some code to understand this approach.

There might also be a  workaround for 3.11+: https://github.com/encode/starlette/pull/1258#issuecomment-1138697483

> For the read request body issue, I would be in favor of keeping that limitation in BaseHTTPMiddleware, and document it more strongly, and I would even look into raising an exception if dispatch calls both request.receive (directly or indirectly) and call_next

I had a proof of concept for this at one point:  https://github.com/encode/starlette/pull/1528

> that is the consequence of call_next hooking itself directly to request.receive, and of there being no attempt at buffering the request body (which, in my opinion, would be a bad idea)

Have you had a chance to look at https://github.com/encode/starlette/pull/1692? My logic here was that `Request.body()` already is expected to store the body in memory, so buffering that should be okay. So by subclassing Request I was able to (1) differentiate `Request.body()` from `Requesr.stream()`, (2) avoid modifying Request itself / only doing this in BaseHTTPMiddleware and (3) only buffering for the downstream middleware / downstream middleware can still consume the stream, modify/replace it, etc. without having to deal with cache invalidation and such.

**Reply by jhominal** [reply]
Oh, it does seem that 3.11 will have a feature that has the potential to solve the issue of context variables bubbling to parent tasks. Interesting. Probably too late to save `BaseHTTPMiddleware`, but interesting!

For the request content, I am of the mind that currently, because of `call_next` using `request.receive` directly, I know that it does not work to consume the body in both `dispatch` and in `call_next` (unless someone writes a new `request.receive` function, but then they are already half-way to implementing an ASGI middleware). Right now, I would prefer warding off that limitation in a deterministic way, e.g. by forbidding call of `call_next` when the body has been consumed by `dispatch`.

I would be extremely wary of allowing `.body()` calls by a middleware to trigger loading of the whole body in memory, because I think that the middleware, by its nature, generally does not have the necessary context to take the decision of "is that content loadable or not", and would lead to an epidemic of "I have a middleware, I will call `.body()` to read the content" completely neutralizing streaming request bodies. 

I think that conceptually, HTTP middleware should handle request and response bodies either in a streaming fashion, or not at all.  And there is a gap here - while there exists a `StreamingResponse` in starlette to handle response bodies in a streaming way, no similar tooling exists for handling of `Request` bodies, as `.receive` is  the authoritative for the request body.

**Comment by jhominal** [comment]
Here is my starting proposal for fixing contextvars: https://github.com/jhominal/starlette/pull/1

It is still a work in progress - in particular, as written, contextvars set in dispatch will not be visible in the endpoint, but I intend to fix that by providing an argument on `call_next` that takes the list of `contextvars` that should be visible to the application.

**Reply by jhominal** [reply]
Now that I reread this merge request, I realize that I am now running the `send` and `receive` callables in the context of `dispatch`... that may be an issue, need to explore that.

**Comment by jhominal** [comment]
Here is my analysis of the `contextvars` issue, and of the impact of `BaseHTTPMiddleware` on them.

Because of the way that `BaseHTTPMiddleware` spawns tasks in order to bridge ASGI with the Request/Response API that it exposes, that means that the user code that is run in a spawned task can be unable to set context variables that can be read in the parent task.

Here are the various pieces of user code that interact with `BaseHTTPMiddleware`, where context variables may be used:

 * Upstream: the caller of `BaseHTTPMiddleware.__call__`;
 * `app`: the downstream application, wrapped by `BaseHTTPMiddleware`;
 *  `dispatch`: the middleware user code;
 * `receive`: the `receive` callable set in `__call__`
 * `send`: the `send` callable set in `__call__`

Here is how the tasks are currently organized in starlette:

 * Main task: Upstream -> `dispatch`/`receive`/`send`
 * Spawned task: (Copy of `dispatch` context) -> `app`

Which leads to this matrix of "who can read context variables set by whom":

| Context variable set by | Upstream | `app` | `dispatch` | `receive` | `send` |
| - | - | - | - | - | - |
| Can be read by Upstream | ✅ | ❌ | ✅ | ✅ | ✅ |
| Can be read by `app` | ✅ | ✅ | ✅ | ❌ | ❌ |
| Can be read by `dispatch` | ✅ | ❌ | ✅ | ✅ | ✅ |
| Can be read by `receive` | ✅ | ❌ | ✅ | ✅ | ✅ |
| Can be read by `send` | ✅ | ❌ | ✅ | ✅ | ✅ |

My main issue with that matrix, is that `app` cannot set variables that will be read upstream - that means that, like in bug #1438, the simple existence of a `BaseHTTPMiddleware`-based middleware in the chain introduces dysfunctions that can only be fixed by the application deployer (by stopping the usage of middlewares that subclass `BaseHTTPMiddleware`) or by the middleware implementer (stop subclassing `BaseHTTPMiddleware`), but not by the victim (the application that sets context variables, or the upstream code that needs to read these context variables).

In order to fix my main issue, I propose the following change in the way that user code is run in various tasks:

 * Main task: Upstream -> `app`
 * Spawned task: (Copy of Upstream context) -> `dispatch`/`receive`/`send`

Such a change was implemented, as a work in progress, in https://github.com/jhominal/starlette/pull/1/ 

That change leads us to *this* matrix:

| Context variable set by | Upstream | `app` | `dispatch` | `receive` | `send` |
| - | - | - | - | - | - |
| Can be read by Upstream | ✅ | ✅ | ❌ | ❌ | ❌ |
| Can be read by `app` | ✅ | ✅ | ❌ | ❌ | ❌ |
| Can be read by `dispatch` | ✅ | ❌ | ✅ | ✅ | ✅ |
| Can be read by `receive` | ✅ | ❌ | ✅ | ✅ | ✅ |
| Can be read by `send` | ✅ | ❌ | ✅ | ✅ | ✅ |

Looking at this matrix, there are two regressions brought by the change:
 * `dispatch` cannot set context variables that can be read Upstream, or by the `app`;
   *  In order to handle that case, I would propose adding a new argument to `call_next`, where `dispatch` can specify a list of `contextvars`, the value of which would be flowed to `app` (and upstream);
 * `receive` and `send` cannot set context variables that will be readable upstream.
   * I have a hard time imagining that there are many `receive` and `send` implementations that set context variables, in the expectation of these context variables being readable everywhere. That is particularly the case for `receive`, which, in many cases, is going to be called from a new task if the application wishes to handle `http.disconnect`;
   * Thus, I would be in favor of accepting that change in behavior;

What does everyone else think?

**Reply by adriangb** [reply]
This is an awesome analysis, thank you very much for working on this!

I agree that the largest problem right now is that `app` can't set context variables which as you say can break the entire app or even some library just by having any BaseHTTPMiddleware in the stack 😓

I also agree that `send` and `receive` are probably not setting or reading context vars, I'm not too worried about changing that, but we would have to note it in the changelog or something.

I think the change for `dispatch` makes sense. I'm really curious if anyone is setting context variables there. I think adding some parameter to track context variables is adding complexity and doesn't change the fact that it would be a breaking change. If we're going to make a breaking change, wouldn't it be more reasonable to ask the (I am guessing small) fraction of users impacted to rewrite their middleware using pure ASGI.

I'm also curious: does this fix the BackgroundTask / cancellation issue?

**Reply by jhominal** [reply]
1. I have realized that the context for `receive` tasks is actually a lot more complex than the matrices I made (`receive` can be called by `app`, `dispatch` (if `dispatch` handles the request body), or by a task spawned by `StreamingResponse.listen_for_disconnect`), and my head is currently exploding because I tried updating the previous post to take all these changes in account... I think we should probably just accept that there cannot be an expectation of `send`/`receive` working correctly regarding them setting `contextvars`, until the Python 3.11 API for `create_task` is usable (a smaller issue exists even without `BaseHTTPMiddleware`);
2. Yes, it is really about giving current users of `BaseHTTPMiddleware` a chance of continuing their usage of `contextvars` in a way that they can keep mostly the same behavior (the compatibility code if we implemented a `flow_contextvars` argument in `call_next` would probably be relatively easy to write with just a bit of `inspect` on `call_next`) without having to completely rewrite their middlewares - but yes, I do not know how many people are actually using context variables and that *would* introduce a bit of additional complexity to `BaseHTTPMiddleware`, without fully solving the issue;
3. The BackgroundTask / cancellation issue is a completely separate issue from context variables... But I did build https://github.com/jhominal/starlette/pull/1 on top of the branch for #1715, and thus https://github.com/jhominal/starlette/pull/1 integrates my fix for the BackgroundTask / cancellation issue - the corresponding tests thus run on my branch;

**Reply by adriangb** [reply]
> 1. I have realized that the context for `receive` tasks is actually a lot more complex than the matrices I made (`receive` can be called by `app`, `dispatch` (if `dispatch` handles the request body), or by a task spawned by `StreamingResponse.listen_for_disconnect`), and my head is currently exploding because I tried updating the previous post to take all these changes in account... I think we should probably just accept that there cannot be an expectation of `send`/`receive` working correctly regarding them setting `contextvars`, until the Python 3.11 API for `create_task` is usable (a smaller issue exists even without `BaseHTTPMiddleware`);

👍 for me at least it would not be a blocker to say that context variables won't work in `receive` and `send`

> 2. Yes, it is really about giving current users of `BaseHTTPMiddleware` a chance of continuing their usage of `contextvars` in a way that they can keep mostly the same behavior (the compatibility code if we implemented a `flow_contextvars` argument in `call_next` would probably be relatively easy to write with just a bit of `inspect` on `call_next`) without having to completely rewrite their middlewares - but yes, I do not know how many people are actually using context variables and that _would_ introduce a bit of additional complexity to `BaseHTTPMiddleware`, without fully solving the issue;

Gotcha. I guess we could do some investigation into public usage or maybe even put in a warning for a couple releases and see if anyone pops up and says please don't break my code.

> 3. The BackgroundTask / cancellation issue is a completely separate issue from context variables... But I did build [reorganize BaseHTTPMiddleware tasks so that app is in the same task as `__call__` jhominal/starlette#1](https://github.com/jhominal/starlette/pull/1) on top of the branch for [Replace task cancellation in `BaseHTTPMiddleware` with `http.disconnect`+`recv_stream.close` #1715](https://github.com/encode/starlette/pull/1715)

Right I understand they are separate issues, I was just confirming if that fix was also included because it looked like it from the source.

--------------------

Since you mention the 3.11+ change to allow passing in your own `Context` instance when you create a task, would it be an option to only fix this for 3.11+? It seems like the least complex solution and I think it would ✅ the entire matrix.  The tradeoff of course is that only users on 3.11+ will benefit, but over time that will become most users, so the cost is time not complexity.

**Reply by jhominal** [reply]
> Since you mention the 3.11+ change to allow passing in your own `Context` instance when you create a task, would it be an option to only fix this for 3.11+? It seems like the least complex solution and I think it would ✅ the entire matrix. The tradeoff of course is that only users on 3.11+ will benefit, but over time that will become most users, so the cost is time not complexity.

Do you mean the fix for "upstream should `contextvars` from `app`" (fixed by https://github.com/jhominal/starlette/pull/1) or "`app` should see `contextvars` from `dispatch`" (introduced by https://github.com/jhominal/starlette/pull/1)?

**Reply by adriangb** [reply]
I think if we applied that change to our current state the matrix would look like this:

| Context variable set by | Upstream | `app` | `dispatch` | `receive` | `send` |
| - | - | - | - | - | - |
| Can be read by Upstream | ✅ |  ✅ | ✅ | ✅ | ✅ |
| Can be read by `app` | ✅ | ✅ | ✅ | ❌ | ❌ |
| Can be read by `dispatch` | ✅ |   ❌ | ✅ | ✅ | ✅ |
| Can be read by `receive` | ✅ | ❌ | ✅ | ✅ | ✅ |
| Can be read by `send` | ✅ | ❌ | ✅ | ✅ | ✅ |

I think it would fix the "contextvar set in app can be read by upstream" issue, which I think is the most important one. I don't think it would fix any of the other issues.

**Reply by jhominal** [reply]
I am a bit confused right now. What change are you talking about that would bring that situation?

As a point of reference:
 * my first matrix covers both the current state on `encode/starlette/master`, and the state after applying #1715 (which I hope will be done). In that case, context variables set by `app` do not flow upstream;
 * my second matrix covers the state after applying https://github.com/jhominal/starlette/pull/1. In that case, context variables set by `dispatch` do not flow either to `app`, or upstream;

As the matrix from your last comment differs from both of mine, I do not know what change would bring that matrix about.

**Reply by adriangb** [reply]
I think that change would look something like this, but I have not tried it yet so who knows if it will work: https://github.com/adriangb/starlette/pull/51/files#diff-931c53548e8908544bd0818ee1a8d5f1ada4f07430a439e9322ab369eb40a8ac

Refs:
#1258
https://github.com/python/cpython/pull/31837

**Reply by adriangb** [reply]
I think that change would look something like this, but I have not tried it yet so who knows if it will work: https://github.com/adriangb/starlette/pull/51/files#diff-931c53548e8908544bd0818ee1a8d5f1ada4f07430a439e9322ab369eb40a8ac

Refs:
#1258
https://github.com/python/cpython/pull/31837

**Reply by jhominal** [reply]
Here are my off the cuff reactions, just looking at your proposed change:

1. Until I saw your proposed change I had thoughtlessly believed that https://github.com/python/cpython/pull/31837 would allow us to run subtasks in *the same context* as the parent task, and of course that is not the case (there is no API to get a reference to the current `Context` object, and a single `Context` object cannot be used to run two functions at a time);
2. Enumerating context keys to detect which context variables have changed is something that I would prefer to avoid - it goes a bit against part of the promise of contextvars, which is that, no matter how many variables you have, you can pop subtasks without incurring a runtime cost (and the underlying data structure is heavily optimized for that purpose);
3. I suspect that there may be a way to emulate the context argument from `create_task` on Python<3.11, by using `Context.run` and hacking at the `__await__` level, but that would have to be confirmed in one way or another;
4. I would still prefer to apply https://github.com/jhominal/starlette/pull/1 before applying a change like yours to synchronize context variables across contexts, because I believe that gives us more options in tuning the "synchronize context vars" strategy;

**Reply by adriangb** [reply]
> 1. Until I saw your proposed change I had thoughtlessly believed that [bpo-46994: Accept explicit contextvars.Context in asyncio create_task() API python/cpython#31837](https://github.com/python/cpython/pull/31837) would allow us to run subtasks in _the same context_ as the parent task, and of course that is not the case (there is no API to get a reference to the current `Context` object, and a single `Context` object cannot be used to run two functions at a time);

Yes, you're right. @agronholm  and I asked about this: https://discuss.python.org/t/back-propagation-of-contextvar-changes-from-worker-threads/.

> 2. Enumerating context keys to detect which context variables have changed is something that I would prefer to avoid - it goes a bit against part of the promise of contextvars, which is that, no matter how many variables you have, you can pop subtasks without incurring a runtime cost (and the underlying data structure is heavily optimized for that purpose);

Theoretically, I agree. Practically though, [Django has had this for years](https://github.com/django/asgiref/blob/22c6767e753007c83a85962e7f460aa6358b2398/asgiref/sync.py#L18-L26) and I haven't heard anyone complain about performance. I also ran a quick benchmark:

```python
from contextvars import ContextVar
from time import time
import anyio

vars = tuple([
    ContextVar[int](str(n))
    for n in range(1_000)
])
for n, var in enumerate(vars):
    var.set(n)

start = time()
for _ in range(1_000):
    for var in vars:
        val = var.get()
        if val % 10 == 0:
            var.set(val)
end = time()
print(end - start)

async def task():
    ...

start = time()
for _ in range(1_000):
    async with anyio.create_task_group() as tg:
        tg.start_soon(task)
end = time()
print(end - start)
```

I got `0.09` and `0.07` respectively. I chose `create_task_group()` as an arbitrary unit of work that BaseHTTPMiddleware is already doing (it's also doing a lot more stuff). So I think in practice the performance impact wouldn't be _that_ much greater than just using BaseHTTPMiddleware in the first place unless someone is using 10,000 context variables _and_ setting hundreds of them for every request, in which case that's a super niche use case and the fix would be to not use BaseHTTPMiddleware (they currently wouldn't even be able to use it, so slow would be an improvement over broken even for this niche use case).

> 4. I would still prefer to apply [reorganize BaseHTTPMiddleware tasks so that app is in the same task as `__call__` jhominal/starlette#1](https://github.com/jhominal/starlette/pull/1) before applying a change like yours to synchronize context variables across contexts, because I believe that gives us more options in tuning the "synchronize context vars" strategy;

I'm open to any solution (and I think yours is pretty amazing), I just want us to compare pros and cons.

**Reply by jhominal** [reply]
> So I think in practice the performance impact wouldn't be that much greater than just using BaseHTTPMiddleware in the first place unless someone is using 10,000 context variables and setting hundreds of them for every request, in which case that's a super niche use case and the fix would be to not use BaseHTTPMiddleware (they currently wouldn't even be able to use it, so slow would be an improvement over broken even for this niche use case).

Very fair. You are right, I should not get too caught up on performance in `BaseHTTPMiddleware`, given how much work it already does in order to be able to provide its interface.

I think that waiting for Python 3.11+ to fix the "`app` cannot set context variables that can be read upstream of a `BaseHTTPMiddleware`" seems a bit late to me. I will think of how to get a fix with "copying context variables" that would work with current APIs (I think that calling `copy_context()` from `coro` *could* work? Or maybe it does not work with the workflows as they are, it does not seem that I can get a handle on the context at the right time.)

**Reply by jhominal** [reply]
I think that, using the principle of "copying context variables", I was able to evolve https://github.com/jhominal/starlette/pull/1 so that:
 * context variables that are set in `dispatch` are available both upstream, and downstream in the app;
 * it works on Python 3.7+, no need to wait for Python 3.11;

There is still a tradeoff in that, in this version of the fix, context var modifications that happen *after* `call_next` is called are not transferred upstream.

In other words, here is the updated matrix of context var visibility, having removed `send`/`receive` from the analysis, and adding a distinction for "`dispatch` before `call_next`" vs "`dispatch` after `call_next`":

##### Before https://github.com/jhominal/starlette/pull/1

| Context variable set by | Upstream | `app` | `dispatch` before `call_next` | `dispatch` after `call_next` |
| - | - | - | - | - |
| Can be read by Upstream | ✅ | ❌ | ✅ | ✅ |
| Can be read by `app` | ✅ | ✅ | ✅ | ❌ |
| Can be read by `dispatch` before `call_next` | ✅ | N/A | ✅ | N/A |
| Can be read by `dispatch` after `call_next`  | ✅ | ❌ | ✅ | ✅ |

##### After https://github.com/jhominal/starlette/pull/1 (both commits)

| Context variable set by | Upstream | `app` | `dispatch` before `call_next` | `dispatch` after `call_next` |
| - | - | - | - | - |
| Can be read by Upstream | ✅ | ✅ | ✅ | ❌ |
| Can be read by `app` | ✅ | ✅ | ✅ | ❌ |
| Can be read by `dispatch` before `call_next` | ✅ | N/A | ✅ | N/A |
| Can be read by `dispatch` after `call_next`  | ✅ | ❌ | ✅ | ✅ |

In other words, we exchange the bug "downstream context variables do not flow upstream" with "context variables from middleware, after `call_next`, do not flow upstream". I think that exchange is more than fair, especially as a middleware can still use mutable objects in the context vars to make changes that will be available to the concurrent tasks.

**Reply by adriangb** [reply]
> I think that exchange is more than fair

💯 agreed. I think that as long as `Upstream` and `app` can communicate we are in a much better place. The issue currently is that if you (Jean) are writing an app and use some library that I (Adrian) wrote that is implemented using `BaseHTTPMiddleware` and that breaks some ContextVar used by your `app`/`Upstream` the only thing you can do is come ask me to change my middleware to accommodate some other library or your private code. With this change, what would break (if anything) is my middleware -> you are now asking me to fix a bug in my middleware. I don't know of _any_ middleware that is setting/getting context variables within `dispatch` and if they exist I think the answer is simple: rewrite it as a pure ASGI middleware.

**Comment by jhominal** [comment]
After reading https://github.com/encode/starlette/issues/1678#issuecomment-1172916042 it appears that the following issue has also been identified in `BaseHTTPMiddleware`, which seems to be that exceptions can be swallowed, due to race conditions that exist around the `app_exc` variable.

I think that fixing that race condition should be possible.

**Reply by adriangb** [reply]
Has anyone looked into this further / worked on it?

**Comment by jhominal** [comment]
I have also continued to think about the issue of processing of request bodies in a `BaseHTTPMiddleware`, and I think the best solution for that issue would be to implement these two action items:

1. Completely preclude the possibility of double consumption by checking that `dispatch` only calls at most one of `request.receive` and `call_next`, and raise an exception if that condition is not met.
2. In order to allow people to process the request (and response) bodies in a streaming fashion, I would propose adding the following lines to `base.py`:

```python
AsyncByteStream = typing.AsyncIterable[bytes]
AsyncByteStreamFunction = typing.Callable[[AsyncByteStream], AsyncByteStream]


class BaseHTTPMiddleware:

    @staticmethod
    def wrap_request_body(request: Request, func: AsyncByteStreamFunction) -> Request:
        wrapped_stream_iterator = func(request.stream()).__aiter__()
        stream_exhausted = False

        async def wrapped_receive() -> Message:
            nonlocal stream_exhausted
            if not stream_exhausted:
                try:
                    chunk = await wrapped_stream_iterator.__anext__()
                    return {"type": "http.request", "body": chunk, "more_body": True}
                except StopAsyncIteration:
                    stream_exhausted = True
                    return {"type": "http.request", "body": b"", "more_body": False}
                except ClientDisconnect:
                    stream_exhausted = True
                    return {"type": "http.disconnect"}
            else:
                return await request.receive()

        return Request(scope=request.scope, receive=wrapped_receive, send=request._send)

    @staticmethod
    def wrap_response_body(response: Response, func: AsyncByteStreamFunction) -> Response:
        if not isinstance(response, StreamingResponse):
            raise ValueError(
                "wrap_response_body can be used only on responses issued by call_next"
            )

        response_with_wrapped_content = StreamingResponse(
            content=func(response.body_iterator),
            status_code=response.status_code,
        )
        response_with_wrapped_content.raw_headers = response.raw_headers
        return response_with_wrapped_content
```

In other words, we could provide helper functions that help users of `BaseHTTPMiddleware` to process the content stream in a streaming fashion, whether in the request or in the response.

**Comment by antonpirker** [comment]
Very interesting discussion!

I built the official Sentry integration for Starlette and a user has now exact the problem of the request freezing because Sentry reads the request and then a `BaseHTTPMiddleware`-based middleware tries to read the request again. 

See: https://github.com/getsentry/sentry-python/issues/1573

Is there any work-around that I could implement to prevent the hanging from happening? 
I can not think of anything that I could do in the Sentry SDK, to allow middlewares that run after Sentry to access the request body with `BaseHTTPMiddleware`s `call_next()` method...

Any pointers or hints are greatly appreciated.

**Reply by adriangb** [reply]
Could you rewrite it as a [pure ASGI middleware](https://github.com/encode/starlette/blob/master/docs/middleware.md#pure-asgi-middleware)?

**Reply by adriangb** [reply]
Sorry looking at the code I realized that the Sentry middleware itself is an ASGI middleware but the user's middleware is not. For the record, it looks like the body is consumed [here](https://github.com/getsentry/sentry-python/blob/fa4f5b03c2d686e1dfb40543d0d099e5391850a9/sentry_sdk/integrations/starlette.py#L487).

**Reply by adriangb** [reply]
I think you have two options here:
1. Buffer the request body and pass this buffer onto downstream apps.
2. Wrap `receive` so that as the downstream application streams the request body you keep track of the size.

(1) is easier to implement in principle. The simplest implementation where you just store the body into memory is trivial but is not suitable for production because you can easily run into memory exhaustion. You could use `SpooledTemporaryFile` or something like that but even then if a user is trying to stream bytes into a database, proxy them to another service, etc. it would be a huge performance hit to write everything to disk as it passes through the app.

(2) has a lot more edge cases from the get go (e.g. what to do if the request body is never consumed by the app or if the app starts consuming it and then crashes) but I think this would scale better to large requests, high concurrency, etc. One issue here is that you won't have the request body size until _after_ the request is done processing (although I assume you batch metadata exports so it's not realtime anyway).

This said, I think trying to get the _real bytes_ from the request body is problematic in general. I don't know if it's possible/useful but I think I would have implemented this metadata in two pieces: (1) the value from the `Content-Length` header (if it exists) and (2) the bytes read by the app (which often, but not always, will be the entire request body). Then solution (2) would be easy to implement because there aren't really any edge cases.

**Reply by antonpirker** [reply]
Thanks for this great input @adriangb 
So this is nothing I can implement in an hour or two :-)
We have to work on the next release of our SDK (going out this week), but afterwards I will have more time to dive into this.

**Reply by adriangb** [reply]
Sounds good, feel free to ping if you have any questions or need examples

**Comment by adriangb** [comment]
Coming back to this. I think the state of the issues and their possible solutions are:
- BackgroundTask cancellation: #1715 and/or #1700 (I think we should do both)
- Reading request bodies: #1692 or #1519 (I think one or the other)
- ContextVars: https://github.com/jhominal/starlette/pull/1 (@jhominal is this ready for a PR to Starlette itself?)
- https://github.com/encode/starlette/issues/1678#issuecomment-1172916042: not sure about this, I don't know that anyone has looked into it.
